extension/src/uris.ts

```diff
@@ -0,0 +1,63 @@
+import * as sourcegraph from 'sourcegraph'
+
+/**
+ * @param textDocumentUri The Sourcegraph text document URI, e.g. `git://github.com/sourcegraph/extensions-client-common?80389224bd48e1e696d5fa11b3ec6fba341c695b#src/schema/graphqlschema.ts`
+ * @returns The root URI for the server, e.g. `https://accesstoken@sourcegraph.com/.api/github.com/sourcegraph/extensions-client-common@80389224bd48e1e696d5fa11b3ec6fba341c695b/-/raw/`. Always has a trailing slash.
+ */
+export function resolveServerRootUri(textDocumentUri: URL): URL {
+    const rootUri = new URL(sourcegraph.internal.sourcegraphURL.toString())
+    // rootUri.username = accessToken
+    rootUri.pathname =
+        '/.api/' + textDocumentUri.host + textDocumentUri.pathname + '@' + textDocumentUri.search.substr(1) + '/-/raw/'
+    return rootUri
+}
+
+/**
+ * @param textDocumentUri The Sourcegraph text document URI like git://github.com/sourcegraph/extensions-client-common?80389224bd48e1e696d5fa11b3ec6fba341c695b#src/schema/graphqlschema.ts
+ * @returns The text document URI for the server, e.g. https://accesstoken@sourcegraph.com/.api/github.com/sourcegraph/extensions-client-common@80389224bd48e1e696d5fa11b3ec6fba341c695b/-/raw/src/schema/graphqlschema.ts
```

Comment:
accesstoken@ will not work, you will need to use `?token=...`

Reply:
No, the PR I closed supported this: https://github.com/sourcegraph/sourcegraph/pull/874

And the new PR examples also clearly call this out: https://github.com/sourcegraph/sourcegraph/pull/800/files#diff-ebbb07fb66365c4b6c3cfd6be0b4916eR89

From the Slack discussion with Beyang and you it sounded like this was purely cosmetic and not needed to me. Why can't you tack on ?token= to your URLs?

Reply:
@beyang's complaint was supporting this for UI routes, but I am using `/.api` here, so I would expect this to work. I explained why on Slack too: https://sourcegraph.slack.com/archives/CCLF4R6EM/p1541724525147200?thread_ts=1541720380.139600&cid=CCLF4R6EM

If I pass a root URI to the language server like `https://sourcegraph.com/.api/github.com/sourcegraph/repo@rev/-/raw/?token=abc123` and then the language server wants to fetch file `test.ts`, it will resolve the URL for `test.ts` relatively to the root URI. The query parameter gets lost in that process because it is tied to the path, not to the host. The userinfo part of the URL is part of the authority (tied to the host). 

Illustrative example:
```ts
new URL('test.ts', 'https://test.com?token=abc').toString() === 'https://test.com/test.ts'
new URL('test.ts', 'https://abc@test.com').toString() === 'https://abc@test.com/test.ts'
```

Support for tokens in the authinfo part is a strict requirement for the whole plan to work.

Reply:
Cool, well that def. was not clear to me that JS/TS has some odd requirement like this. Apologies for not inquiring more about this earlier. I'll send a new change to support this later tonight and Beyang can review whenever he has time.